babel/src/main/codegen/includes/parserImpls.ftl

```diff
@@ -692,3 +692,80 @@ SqlNode DateTimeTerm() :
         )
     )
 }
+
+// This excludes CompoundIdentifier() as a data type.
+SqlDataTypeSpec DataTypeAlternativeCastSyntax() :
+{
+    SqlTypeNameSpec typeName;
+    final Span s;
+}
+{
+    typeName = TypeNameAlternativeCastSyntax() {
+        s = span();
+    }
+    (
+        typeName = CollectionsTypeName(typeName)
+    )*
+    {
+        return new SqlDataTypeSpec(
+            typeName,
+            s.end(this));
+    }
+}
+
+// This excludes CompoundIdentifier() as a type name that's found in the
+// original TypeName() function. Custom data types can be parsed
+// in parser.dataTypeParserMethods.
+SqlTypeNameSpec TypeNameAlternativeCastSyntax() :
```

Comment:
This looks like it shares a lot of logic in common with the existing TypeName() function in Parser.jj. Would another approach be to change that function to take an argument indicating whether a compound identifier is valid?

Then we can call that function from here with the argument set to false. And the existing places that invoke it can call with that argument set to true.

WDYT?

Reply:
Yeah that would be a good idea to remove some of this copied code. If we had TypeName with a parameter, would we have still copy the logic in DataType into our custom parsing method? Then the DataType method would just call TypeName with true for compound identifiers?

Reply:
Or did you mean adding a parameter to DataType() and passing that down to TypeName(), then just updating the other DataType calls?

Reply:
Exactly.

Reply:
Cool that sounds like it'll make this a lot cleaner

Reply:
Actually I'm not sure how to use a parameter like this to modify the parsing logic because I don't think you can call parsing logic like TypeName() inside a java if statement. Is there some other way to reference this variable inside parsing logic?